prow/cmd/initupload/main.go

```diff
@@ -0,0 +1,180 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// initupload parses the logs from the clonerefs
+// container and determines if that container was
+// successful or not. Using that information, this
+// container uploads started.json and if necessary
+// finished.json as well.
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/option"
+
+	"k8s.io/test-infra/prow/pjutil"
+	"k8s.io/test-infra/prow/pod-utils/clone"
+	"k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+var (
+	cloneLog = flag.String("clone-log", "", "path to the output file for the cloning step")
+
+	subDir = flag.String("sub-dir", "", "Optional sub-directory of the job's path to which artifacts are uploaded")
+
+	pathStrategy = flag.String("path-strategy", pathStrategyExplicit, "how to encode org and repo into GCS paths")
+	defaultOrg   = flag.String("default-org", "", "optional default org for GCS path encoding")
+	defaultRepo  = flag.String("default-repo", "", "optional default repo for GCS path encoding")
+
+	gcsBucket          = flag.String("gcs-bucket", "", "GCS bucket to upload into")
+	gceCredentialsFile = flag.String("gcs-credentials-file", "", "file where Google Cloud authentication credentials are stored")
+	dryRun             = flag.Bool("dry-run", true, "do not interact with GCS")
+)
+
+type pathStrategyType string
+
+const (
+	pathStrategyLegacy   pathStrategyType = "legacy"
+	pathStrategyExplicit                  = "explicit"
+	pathStrategySingle                    = "single"
+)
+
+func main() {
+	flag.Parse()
+
+	if !*dryRun {
+		if *gcsBucket != "" {
+			logrus.Fatal("No GCS bucket specified")
+		}
+
+		if *gceCredentialsFile != "" {
+			logrus.Fatal("No GCE credentials specified")
+		}
+	}
+
+	if err := validatePathOptions(pathStrategy, defaultOrg, defaultRepo); err != nil {
+		logrus.Fatalf("Invalid path options: %v", err)
+	}
+
+	var builder gcs.RepoPathBuilder
+	switch pathStrategyType(*pathStrategy) {
+	case pathStrategyExplicit:
+		builder = gcs.NewExplicitRepoPathBuilder()
+	case pathStrategyLegacy:
+		builder = gcs.NewLegacyRepoPathBuilder(*defaultOrg, *defaultRepo)
+	case pathStrategySingle:
+		builder = gcs.NewSingleDefaultRepoPathBuilder(*defaultOrg, *defaultRepo)
+	}
+
+	spec, err := pjutil.ResolveSpecFromEnv()
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not resolve job spec")
+	}
+
+	gcsPath := gcs.PathForSpec(spec, builder)
+	if *subDir != "" {
+		gcsPath = path.Join(gcsPath, *subDir)
+	}
+
+	var cloneRecords []clone.Record
+	data, err := ioutil.ReadFile(*cloneLog)
```

Comment:
Validate that `*cloneLog != ""` in case the `--clone-log` flag is mistakenly not provided.